File: Data Structures & Algorithms/balanced-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def dfs(root):
            if root is None:
                return [0, True]
            
            left = dfs(root.left)
            right = dfs(root.right)

            if left[1] and right[1]:
                if abs(left[0] - right[0]) <= 1:
                    return [max(left[0], right[0]) + 1, True]
                else:
                    return [max(left[0], right[0]) + 1, False]
            else:
                return [max(left[0], right[0]) + 1, False]

        return dfs(root)[1]

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        if (preorder.empty() or inorder.empty()) return nullptr;

        TreeNode* root = new TreeNode(preorder[0]);

        auto it = find(inorder.begin(), inorder.end(), preorder[0]);
        int index = distance(inorder.begin(), it);

        vector<int> preleft(preorder.begin() + 1, preorder.begin() + index + 1);
        vector<int> inleft(inorder.begin(), inorder.begin() + index);   
        root->left = buildTree(preleft, inleft);

        vector<int> preright(preorder.begin() + index + 1, preorder.end());
        vector<int> inright(inorder.begin() + index + 1, inorder.end());   
        root->right = buildTree(preright, inright);

        return root;
    }
};
